src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp

```diff
@@ -840,54 +789,44 @@ SubCmdProgram::execute(const SubCmdOptions& _options) const
   if (deviceCollection.size() == 0) 
     throw std::runtime_error("No devices found.");
 
-  // Get the device
-  auto & workingDevice = deviceCollection[0];
-
-  // TODO: Added mutually exclusive code for image, update, and revert-to-golden action.
-
-  if (!image.empty()) {
-
-    // image is a sub-option of update
-    if (update.empty())
-      throw xrt_core::error("Usage: xbmgmt program --device='0000:00:00.0' --base --image='/path/to/flash_image' OR shell_name");
-
-    // We support up to 2 flash images 
-    if (image.size() == 1)
-      auto_flash(workingDevice, flashType, image.front());
-    
-    if (image.size() == 2) {
-      std::cout << "CAUTION! Force flashing the platform on the device without any checks." <<
-                   "Please make sure that the correct information is passed in." << std::endl;
-      auto image_paths = find_flash_image_paths(image);
-      update_shell(workingDevice->get_device_id(), flashType, image_paths.front(), (image_paths.size() == 2 ? image_paths[1]: ""));
-    }
-    
-    if (image.size() > 2)
-      throw xrt_core::error("Please specify either 1 or 2 flash images");
-    return;
-  }
-
   if (!update.empty()) {
-    XBU::verbose("Sub command: --base");
-    XBUtilities::sudo_or_throw("Root privileges are required to update the devices flash image");
-    if (update.compare("all") == 0)
-      // Note: To get around a bug in the SC flashing code base,
-      //       auto_flash will clear the collection. This code need to be refactored and clean up.
-      auto_flash( workingDevice );
-    else {
-      if (update.compare("flash") == 0)
-        throw xrt_core::error("Platform only update is not supported");
-
-      if (update.compare("sc") == 0)
-        throw xrt_core::error("SC only update is not supported");
-       
-      throw xrt_core::error("Please specify a valid value");
+    // Get a list of images known exist
+    const auto validated_images = find_flash_image_paths(image);
+    if (validated_images.size() <= 2) {
+      // Get the device
+      auto & workingDevice = deviceCollection[0];
+      XBU::verbose("Sub command: --base");
+      XBUtilities::sudo_or_throw("Root privileges are required to update the devices flash image");
+      if (update.compare("all") == 0) {
+        if (validated_images.empty())
+          auto_flash(workingDevice);
+        else if(boost::filesystem::extension(validated_images[0]).compare(".xsabin") == 0)
```

Comment:
The first means that no image after the find_image_flash_paths call was valid (it doesn't exist or was not given in the first place). So we assume auto flash with no image specified. It would probably be better to write image.empty() instead. Second means we have a file that exists, but, the extension may not be xsabin. The find_flash_image_path function returns a file if it exists or matches a shell.

Reply:
Oh, so the first case is where we do the auto-update, i.e., find the file on disk. and 2nd is where the user specifies the `--image` option. Right?

Reply:
I guess if there are some questions about the logic here, are not comments missing in the code?

Reply:
Yes, some comments will be helpful. In fact, the confusing part for me was the function name `auto_flash`. Auto flash suggests that we look up the xsabin for the user and flash it. Thus the `else if` with passing n the image doesn't qualify as an auto_flash. Ideally, the function should be renamed to `flash_shell_and_SC` and have a comment in L801 indicating that we are doing an auto flash

Reply:
Code is much simpler and has comments now!